PlayLoop/PlayerInteractionHandler.js

```diff
@@ -0,0 +1,201 @@
+        // should probably merge updateControl stuff into Controller and read from there instead?
+        let buttonDown = Controller[id];
+        this.justPressed[id] = buttonDown && !this.pressed[id];
+        this.pressed[id] = buttonDown;
+    }
+
+    static handleInput() {
+        // should probably merge updateControl stuff into Controller and read from there instead?
+        // controls need to be updated every frame
+        this.updateControl('jump');
+        this.updateControl('up');
+        this.updateControl('down');
+
+        if (!this.currentTarget) {
+            return;
+        }
+        if (this.blockers.length > 0) {
+            this.recentlyBlocked = true;
+            return;
+        }
+
+        if (this.recentlyBlocked) {
+            // wait an extra frame before interacting if recently blocked
+            this.recentlyBlocked = false;
+            return;
+        }
+
+        if (this.currentTarget.type === ObjectTypes.NPC) {
+            if (this.justPressed.jump) {
+                this.currentTarget.startDialogue();
+            }
+        }
+        if (this.currentTarget.type === ObjectTypes.DOOR) {
+            let triggerType = this.currentTarget.activationTrigger;
+            let cueEntrance = false;
+
+            if (triggerType === DOOR_TRIGGERS.AUTOMATIC) {
+                cueEntrance = true;
+            }
+            else if (triggerType === DOOR_TRIGGERS.UP) {
+                cueEntrance = this.justPressed.up;
+            }
+            else if (triggerType === DOOR_TRIGGERS.DOWN) {
+                cueEntrance = this.justPressed.down;
+            }
+            
+            if (cueEntrance) {
+                this.currentTarget.sendPlayerThroughExit();
+            }
+        }
+    }
+
+    static updatePlayerTarget(player) {
+        if (player.death) {
+            this.lastTarget = this.currentTarget;
+            this.currentTarget = null;
+            return;
+        }
+
+        let playerX = player.x;
+        let playerY = player.y;
+        let minDist = Infinity;
+        let nearestTarget = null;
+
+        this.registeredTargets.forEach((target) => {
+            let xDist = target.x - playerX;
+            let yDist = target.y - playerY;
+            let dist = Math.sqrt((xDist * xDist) + (yDist * yDist));
+            if (dist < minDist) {
+                minDist = dist;
+                nearestTarget = target;
+            }
+        });
+
+        this.lastTarget = this.currentTarget;
+        this.currentTarget = nearestTarget;
+    }
+
+    static updateOverlay() {
+        let target = this.currentTarget;
+
+        if (!target) {
+            // no target means no arrow to draw
+            return;
+        }
+
+        if (this.blockers.length > 0) {
+            // if currently blocked from interaction, don't show the arrow
+            return;
+        }
+
+        // frame-based blink timing copied from orignal NPC code. might want timestamps instead?
```

Comment:
No timestamps pls, because fps taken lag into consideration

Reply:
Ah interesting! Thanks! The comment there was mostly a note-to-self about timing readability — maybe using a constant or two there could be helpful instead.